src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,172 @@
+        category_information = tagging_category_api.get(target_category_id)
+        return true if category_information.value.cardinality == "SINGLE"
+        false
+      end
+
+      def create_tag_hash(vm_config_tags)
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+
+          if vm_config_tag["category"].nil?
+            logger.warn( "Missing category content, skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["category"].empty?
+            logger.warn( "Empty category found in cloud config ,skip processing this category-tag pair." )
+            next
+          end
+          if vm_config_tag["tag"].nil?
+            logger.warn( "Missing tag in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          if vm_config_tag["tag"].empty?
+            logger.warn( "Empty tag found in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
+        end
+        tag_hash
+      end
+
+      def attach_single_tag(vm_mob_id, tag_id )
+        tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttach.new
+        tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        tag_assoc_info.object_id.id = vm_mob_id
+        tag_assoc_info.object_id.type = 'VirtualMachine'
+        tag_assoc_api.attach(tag_id,tag_assoc_info)
+      end
+
+      def attach_multi_tags(vm_mob_id, tag_ids, category_name, target_category_id)
+        single_cardinality = single_category_cardinality?(target_category_id)
+        if single_cardinality
+          logger.warn( "Tag category '#{category_name}' does not support 'Multiple Cardinality', skip tagging this category to vm." )
+          return
+        end
+        multi_tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttachMultipleTagsToObject.new
+        multi_tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        multi_tag_assoc_info.object_id.id = vm_mob_id
+        multi_tag_assoc_info.object_id.type = 'VirtualMachine'
+        multi_tag_assoc_info.tag_ids = tag_ids
+        tag_assoc_api.attach_multiple_tags_to_object(multi_tag_assoc_info)
+      end
+
+      def attach_tags(vm_mob_id, vm_config_tags, vm_config_name)
+        tag_hash = create_tag_hash(vm_config_tags)
+        if tag_hash.empty?
+          logger.warn( "No valid category-tag pair in cloud config, finish attaching tags." )
+          return
+        end
+        category_ids = tagging_category_api.list.value
+        if category_ids.empty?
+          logger.warn( "No category exist on vCenter, skip attaching tags." )
+          return
+        end
+
+        tag_hash.keys.each do |category_name|
+          begin
```

Comment:
This `begin ... end` is unnecessary. `do ... end` forms an implicit `begin .. end` block since Ruby 2.3.

Reply:
Hi @ktchen14, I am using a begin rescue end here to catch exceptions, thus we need to keep  begin, right ?

Reply:
Oh never mind. This feature is only available in Ruby 2.5. Basically in Ruby 2.5 and later you can do:

```ruby
list.each do |i|
  ...
rescue => e
  ...
end
```

But it doesn't work in Ruby 2.4 which is what we're on.

Reply:
Yes, when I try to remove begin and end, RubyMine shows message `rescue/else./ensure inside do blocks  require at least Ruby 2.5 `.